### PR TITLE
feat(anonymizer): handle Managed Agents SSE event types [closes #107]

### LINKED ISSUE #107: monitoring(api-change): Anthropic Managed Agents API introduces unhandled SSE event types
## Critical Finding

**Severity**: Critical — New streaming event types on api.anthropic.com that may bypass the proxy's deanonymization pipeline
**Source**: https://docs.anthropic.com/en/release-notes/overview
**Date published**: 2026-04-08
**Deadline**: Live now (public beta, activated via `anthropic-beta: managed-agents-2026-04-01` header)

### Impact on Proxy

Anthropic's Managed Agents API (public beta) introduces agent-specific SSE streaming events on `api.anthropic.com`. When a client sends the `anthropic-beta: managed-agents-2026-04-01` header, agent endpoints emit new event types that the proxy's Anthropic SSE event parser was not designed to handle. If these events contain text content with PII tokens, the deanonymization step may fail silently — PII tokens could pass through to the client without being replaced with original values.

### Affected Code Paths

- Anthropic provider-specific SSE event parser (streaming response handler)
- SSE event type dispatch/registry
- Deanonymization pipeline for streaming responses
- Beta header passthrough logic

### Suggested Response

1. Inventory the new Managed Agents SSE event types from the Anthropic streaming docs
2. Update the Anthropic event parser to recognize and process agent-specific events
3. Verify that all text content fields in agent events pass through the deanonymization pipeline
4. Add integration tests for agent streaming responses containing PII tokens
5. Consider whether the `anthropic-beta` header should trigger any special proxy behavior

### References

- Monthly digest: https://github.com/laplaque/ai-anonymizing-proxy/discussions/106
- Wiki context: [Provider API Contracts](https://github.com/laplaque/ai-anonymizing-proxy/wiki/Provider-API-Contracts)

### PR DESCRIPTION
**Pull request by iamclaude697**
## What

Anthropic's Managed Agents API (beta: `managed-agents-2026-04-01`) emits agent-level SSE events with a different JSON schema than the standard Messages API. Without explicit handling, PII tokens in `content[].text` fields of agent events could leak through if they match JSON structure rather than raw text.

## Changes

- Add `agentContentBlock` / `agentEventEnvelope` structs to parse agent events
- Route `agent.*` event types to targeted handlers in `ProcessDataPayload`
- `agent.message` / `agent.tool_result` / `agent.mcp_tool_result`: extract and replace PII tokens in `content[].text` fields, re-serialize preserving all other fields
- `agent.tool_use` / `agent.custom_tool_use` / `agent.mcp_tool_use`: replace tokens in serialized `input` field
- `agent.thinking` and unknown `agent.*` events: raw replacement passthrough (defense in depth)
- 14 new tests covering all agent event types, mixed streams, and regressions

## Quality Gates

- [x] `make check` passed: `All checks passed.`
- [x] `go test -race ./...` passed
- [x] gosec: 0 issues (28 files scanned)
- [x] govulncheck: no vulnerabilities
- [x] Architecture invariants maintained

## Test Plan

- Each agent event type tested with PII tokens in text fields
- Empty content arrays pass through without mutation
- Unknown future agent event types get raw replacement fallback
- Mixed stream (standard content_block_delta + agent events) works correctly
- Regression tests for existing text_delta, thinking_delta, input_json_delta

## Linked Issues

Closes #107

### CODE REVIEW

File: internal/anonymizer/streaming_anthropic.go [hunk 1]
```diff
@@ -130,6 +151,104 @@ func (a *anthropicDeanonymizer) processJSONDelta(envelope *sseEnvelope) {
...
+		return true
+	}
+
+	// Re-serialize from a raw map to preserve all fields we didn't parse.
+	var raw map[string]json.RawMessage
+	if err := json.Unmarshal(payload, &raw); err != nil {
+		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
+		return true
+	}
+	contentBytes, _ := json.Marshal(agent.Content)
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_anthropic.go]: **HIGH — Content block re-serialization drops unknown fields**

`agentContentBlock` only captures `type` and `text`:

```go
type agentContentBlock struct {
    Type string `json:"type"`
    Text string `json:"text"`
}
```

When an `agent.message` or `agent.tool_result` event contains a mix of content block types (e.g., a `text` block with PII alongside an `image` or `tool_use` block), the non-text blocks are parsed into `agentContentBlock` — silently dropping all fields except `type` and `text`. Then `json.Marshal(agent.Content)` re-serializes the entire array, producing broken blocks like `{"type":"image","text":""}` instead of the original `{"type":"image","source":{...}}`.

This only triggers when `replaced == true` (i.e., at least one text block had a PII token), but when it does, it corrupts the entire content array.

**Suggested fix:** Instead of re-serializing parsed structs, do targeted replacement on raw JSON content blocks. For example:

```go
// Parse content as raw JSON array, only modify text blocks in-place.
var rawContent []json.RawMessage
if err := json.Unmarshal(raw["content"], &rawContent); err != nil {
    // fallback
}
for i, rc := range rawContent {
    var block struct{ Type, Text string }
    if json.Unmarshal(rc, &block) == nil && block.Type == "text" && block.Text != "" {
        replaced := a.opts.replacer.Replace(block.Text)
        if replaced != block.Text {
            // patch only the "text" field in this raw block
            var m map[string]json.RawMessage
            json.Unmarshal(rc, &m)
            m["text"], _ = json.Marshal(replaced)
            rawContent[i], _ = json.Marshal(m)
        }
    }
}
raw["content"], _ = json.Marshal(rawContent)
```

This preserves all fields on every content block type while only modifying the `text` value on text blocks that contain PII tokens.

File: internal/anonymizer/streaming_anthropic.go [hunk 2]
```diff
@@ -130,6 +151,104 @@ func (a *anthropicDeanonymizer) processJSONDelta(envelope *sseEnvelope) {
...
+		return false
+	}
+
+	switch agent.Type {
+	case "agent.message", "agent.tool_result", "agent.mcp_tool_result":
+		return a.processAgentContentEvent(payload, &agent)
+	case "agent.tool_use", "agent.custom_tool_use", "agent.mcp_tool_use":
+		return a.processAgentInputEvent(payload, &agent)
+	default:
+		// agent.thinking, session.*, span.*, etc. — no text to accumulate.
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_anthropic.go]: **MEDIUM — Misleading comment**

`session.*` and `span.*` events don't start with `agent.` — they can never reach this code path because of the `strings.HasPrefix(envelope.Type, "agent.")` guard in `ProcessDataPayload`. Only `agent.thinking` and other unknown `agent.*` subtypes land here.

```suggestion
		// agent.thinking and other unknown agent.* subtypes — raw replacement passthrough.
```

File: internal/anonymizer/streaming_anthropic_test.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	tokenMap := map[string]string{token: original}
+
+	prefix := strings.Repeat("a", tokenSuffixLen+10)
+	sseInput := makeSSEJsonDelta(prefix+token) + "\n"
+
+	got := readStreamResult(t, sseInput, tokenMap)
+	if !strings.Contains(got, original) {
+		t.Errorf("input_json_delta token not replaced:\n%s", got)
+	}
+}
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_anthropic_test.go:423]: **MEDIUM — Missing test for mixed content block types**

The tests only exercise content arrays where every block is `type: "text"`. There's no test with a mix of text and non-text blocks (e.g., `[{type:"text", text:"PII_TOKEN"}, {type:"image", source:{...}}]`), which is exactly the scenario that triggers the data loss bug above.

Adding a test like this would both document the expected behavior and catch regressions:

```go
func TestAgentMessageMixedContentBlockTypes(t *testing.T) {
    token := "[PII_EMAIL_c160f8cc4b2e1a3d]"
    original := "earl@example.com"
    tokenMap := map[string]string{token: original}

    event := map[string]any{
        "type": "agent.message",
        "id":   "evt_mixed",
        "content": []any{
            map[string]any{"type": "text", "text": "Hello " + token},
            map[string]any{"type": "image", "source": map[string]any{
                "type":       "base64",
                "media_type": "image/png",
                "data":       "iVBOR...",
            }},
        },
    }
    b, _ := json.Marshal(event)
    sseInput := "data: " + string(b) + "\n"
    got := readStreamResult(t, sseInput+"\n", tokenMap)

    // Text block should have PII replaced
    if strings.Contains(got, token) {
        t.Errorf("token not replaced:\n%s", got)
    }
    // Image block must retain its source field
    if !strings.Contains(got, "iVBOR") {
        t.Errorf("image block source data lost during re-serialization:\n%s", got)
    }
}
```

**Review by iamclaude697 — COMMENTED** [review]
## PR Review: feat(anonymizer): handle Managed Agents SSE event types [closes #107]

### Pipeline status
PASS — All 8 checks green (Lint, Build, Test, Security Scan, Benchmark, CodeQL, Analyze).

### Test coverage verdict
PASS — 14 new tests cover all agent event types, empty content, unknown events, mixed streams, and regression cases for existing delta handling. One gap noted: no test for mixed content block types (text + image in same array).

### Findings

| # | Severity | File | Finding |
|---|----------|------|---------|
| 1 | **HIGH** | `streaming_anthropic.go:187-210` | `agentContentBlock` struct only captures `type`+`text`. Re-serialization via `json.Marshal(agent.Content)` drops all other fields from non-text content blocks (image, tool_use, etc.) when a text block in the same array triggers replacement. |
| 2 | **MEDIUM** | `streaming_anthropic.go:171` | Comment mentions `session.*, span.*` but those don't start with `agent.` and can't reach this path. |
| 3 | **MEDIUM** | `streaming_anthropic_test.go:336` | No test for mixed content block types — the exact scenario that exposes finding #1. |

### What's good

- Clean dispatch architecture: `processAgentEvent` → type-specific handlers follows the existing pattern well.
- The raw-map re-serialization approach for preserving unknown envelope fields is solid — just needs to extend one level deeper into content blocks.
- Defense-in-depth default case for unknown `agent.*` events is the right call.
- Test coverage is thorough for the happy path and edge cases.

### Path to Approval

1. **Fix content block re-serialization** (finding #1) — use `[]json.RawMessage` to parse content blocks and do targeted text-field patching, preserving all other fields on every block type. See inline comment for a suggested approach.
2. **Fix misleading comment** (finding #2) — minor, one-line change.
3. **Add mixed-content-type test** (finding #3) — validates the fix for #1.

### Verdict
**Not ready to merge** — the content block data loss is a correctness issue that could silently corrupt non-text content in agent events. The fix is straightforward and contained to `processAgentContentEvent`.

File: internal/anonymizer/streaming_anthropic.go (hunk 1 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go]: **HIGH — Content block re-serialization drops unknown fields**

`agentContentBlock` only captures `type` and `text`:

```go
type agentContentBlock struct {
    Type string `json:"type"`
    Text string `json:"text"`
}
```

When an `agent.message` or `agent.tool_result` event contains a mix of content block types (e.g., a `text` block with PII alongside an `image` or `tool_use` block), the non-text blocks are parsed into `agentContentBlock` — silently dropping all fields except `type` and `text`. Then `json.Marshal(agent.Content)` re-serializes the entire array, producing broken blocks like `{"type":"image","text":""}` instead of the original `{"type":"image","source":{...}}`.

This only triggers when `replaced == true` (i.e., at least one text block had a PII token), but when it does, it corrupts the entire content array.

**Suggested fix:** Instead of re-serializing parsed structs, do targeted replacement on raw JSON content blocks:

```go
var rawContent []json.RawMessage
if err := json.Unmarshal(raw["content"], &rawContent); err \!= nil {
    // fallback
}
for i, rc := range rawContent {
    var block struct{ Type, Text string }
    if json.Unmarshal(rc, &block) == nil && block.Type == "text" && block.Text \!= "" {
        newText := a.opts.replacer.Replace(block.Text)
        if newText \!= block.Text {
            var m map[string]json.RawMessage
            json.Unmarshal(rc, &m)
            m["text"], _ = json.Marshal(newText)
            rawContent[i], _ = json.Marshal(m)
        }
    }
}
raw["content"], _ = json.Marshal(rawContent)
```

This preserves all fields on every content block type while only modifying the `text` value on text blocks that contain PII tokens.

**Review by laplaque — COMMENTED** [review]
## PR Review: feat(anonymizer): handle Managed Agents SSE event types [closes #107]

### Pipeline status
PASS — All 8 checks green (Lint, Build, Test, Security Scan, Benchmark, CodeQL, Analyze).

### Test coverage verdict
PASS — 14 new tests cover all agent event types, empty content, unknown events, mixed streams, and regression cases for existing delta handling. One gap noted: no test for mixed content block types (text + image in same array).

### Findings

| # | Severity | File | Finding |
|---|----------|------|---------|
| 1 | **HIGH** | `streaming_anthropic.go:187-210` | `agentContentBlock` struct only captures `type`+`text`. Re-serialization via `json.Marshal(agent.Content)` drops all other fields from non-text content blocks (image, tool_use, etc.) when a text block in the same array triggers replacement. |
| 2 | **MEDIUM** | `streaming_anthropic.go:171` | Comment mentions `session.*, span.*` but those don't start with `agent.` and can't reach this path. |
| 3 | **MEDIUM** | `streaming_anthropic_test.go:336` | No test for mixed content block types — the exact scenario that exposes finding #1. |

### What's good

- Clean dispatch architecture: `processAgentEvent` → type-specific handlers follows the existing pattern well.
- The raw-map re-serialization approach for preserving unknown envelope fields is solid — just needs to extend one level deeper into content blocks.
- Defense-in-depth default case for unknown `agent.*` events is the right call.
- Test coverage is thorough for the happy path and edge cases.

### Path to Approval

1. **Fix content block re-serialization** (finding #1) — use `[]json.RawMessage` to parse content blocks and do targeted text-field patching, preserving all other fields on every block type.
2. **Fix misleading comment** (finding #2) — minor, one-line change.
3. **Add mixed-content-type test** (finding #3) — validates the fix for #1.

### Verdict
**Not ready to merge** — the content block data loss is a correctness issue that could silently corrupt non-text content in agent events. The fix is straightforward and contained to `processAgentContentEvent`.

File: internal/anonymizer/streaming_anthropic.go (hunk 2 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go]: **MEDIUM — Misleading comment**

`session.*` and `span.*` events don't start with `agent.` — they can never reach this code path because of the `strings.HasPrefix(envelope.Type, "agent.")` guard in `ProcessDataPayload`. Only `agent.thinking` and other unknown `agent.*` subtypes land here.

```suggestion
		// agent.thinking and other unknown agent.* subtypes — raw replacement passthrough.
```

File: internal/anonymizer/streaming_anthropic_test.go (hunk 3 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic_test.go:423]: **MEDIUM — Missing test for mixed content block types**

The tests only exercise content arrays where every block is `type: "text"`. There's no test with a mix of text and non-text blocks (e.g., `[{type:"text", text:"PII_TOKEN"}, {type:"image", source:{...}}]`), which is exactly the scenario that triggers the data loss bug above.

Adding a test like this would both document the expected behavior and catch regressions:

```go
func TestAgentMessageMixedContentBlockTypes(t *testing.T) {
    token := "[PII_EMAIL_c160f8cc4b2e1a3d]"
    original := "earl@example.com"
    tokenMap := map[string]string{token: original}

    event := map[string]any{
        "type": "agent.message",
        "id":   "evt_mixed",
        "content": []any{
            map[string]any{"type": "text", "text": "Hello " + token},
            map[string]any{"type": "image", "source": map[string]any{
                "type":       "base64",
                "media_type": "image/png",
                "data":       "iVBOR...",
            }},
        },
    }
    b, _ := json.Marshal(event)
    sseInput := "data: " + string(b) + "\n"
    got := readStreamResult(t, sseInput+"\n", tokenMap)

    if strings.Contains(got, token) {
        t.Errorf("token not replaced:\n%s", got)
    }
    if \!strings.Contains(got, "iVBOR") {
        t.Errorf("image block source data lost during re-serialization:\n%s", got)
    }
}
```

File: internal/anonymizer/streaming_anthropic.go (hunk 1 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go]: **HIGH** Content block re-serialization drops unknown fields. json.Marshal(agent.Content) only serializes Type+Text, losing all other fields from non-text content blocks (image source data, etc.) when replacement triggers re-serialization. Fix: use []json.RawMessage and patch text fields in-place.

File: internal/anonymizer/streaming_anthropic.go (hunk 2 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go]: **MEDIUM** session.* and span.* can never reach this path (HasPrefix "agent." guard). Only agent.thinking and unknown agent.* subtypes land here.

File: internal/anonymizer/streaming_anthropic_test.go (hunk 3 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic_test.go:423]: **MEDIUM** No test for mixed content block types (text + image in same array) which is the scenario that exposes the data loss bug.

**Review by laplaque — CHANGES_REQUESTED** [review]
Supersedes earlier COMMENT review. See inline comments for findings.

## PR Review: feat(anonymizer): handle Managed Agents SSE event types

### Pipeline status
PASS

### Findings

| # | Severity | File | Finding |
|---|----------|------|---------|
| 1 | **HIGH** | streaming_anthropic.go:187-210 | Content block re-serialization drops unknown fields from non-text blocks |
| 2 | **MEDIUM** | streaming_anthropic.go:171 | Misleading comment references unreachable event types |
| 3 | **MEDIUM** | streaming_anthropic_test.go:336 | No test for mixed content block types |

### Path to Approval
1. Fix content block re-serialization using raw JSON patching
2. Fix misleading comment
3. Add mixed-content-type test

### Verdict
**Not ready to merge**

File: internal/anonymizer/streaming_anthropic.go [hunk 4]
```diff
@@ -130,6 +151,104 @@ func (a *anthropicDeanonymizer) processJSONDelta(envelope *sseEnvelope) {
...
+func (a *anthropicDeanonymizer) processAgentInputEvent(payload []byte, agent *agentEventEnvelope) bool {
+	if len(agent.Input) == 0 {
+		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
+		return true
+	}
+
+	original := string(agent.Input)
+	replaced := a.opts.replacer.Replace(original)
+	if replaced == original {
+		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
```
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go:271]: **MEDIUM:** `processAgentInputEvent` does raw string replacement on `json.RawMessage` content, while `processAgentContentEvent` properly deserializes text fields and re-serializes via `json.Marshal` (which handles JSON escaping).

If a stored PII original value contains JSON-special characters (`"`, `\`, newlines), the raw replacement here could produce invalid JSON output. Example: token `[PII_NAME_abc123...]` maps to original `O"Brien` — replacing inside `{"query":"find [PII_NAME_abc123...]"}` yields `{"query":"find O"Brien"}` (broken).

This is the same pattern used in the existing `default` case (and likely throughout the codebase for raw passthrough), so it's not a regression — but the inconsistency between the two approaches in adjacent functions is worth noting.

**Suggested improvement** (non-blocking): parse `input` into a `map[string]any`, walk string values with the replacer, and re-serialize. This matches the safety of the content-event path.

File: internal/anonymizer/streaming_anthropic_test.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+		t.Errorf("agent.custom_tool_use still contains token:\n%s", got)
+	}
+}
+
+func TestAgentThinkingPassthrough(t *testing.T) {
+	tokenMap := map[string]string{}
+
+	sseInput := makeAgentThinking()
+	got := readStreamResult(t, sseInput+"\n", tokenMap)
+
```
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic_test.go:297]: **LOW:** `agent.mcp_tool_use` is handled by `processAgentInputEvent` alongside `agent.tool_use` and `agent.custom_tool_use`, but only the latter two have dedicated tests. Consider adding a `makeAgentMCPToolUse` helper and a `TestAgentMCPToolUseInputReplacement` for completeness — it's the same code path but confirms the dispatch routing.

**Review by laplaque — COMMENTED** [review]
## PR Review: feat(anonymizer): handle Managed Agents SSE event types [closes #107]

**Scope:** 455 additions, 0 deletions, 2 files.

### Pipeline status

PASS — all 8 checks green (Lint, Test, Benchmark, Build, Security Scan, CodeQL, Analyze (go), Analyze (actions)).

### Test coverage verdict

PASS — 14 new tests covering all dispatched event types, empty-content passthrough, no-mutation when tokens absent, unknown event fallback, mixed stream interleaving, and regression tests for existing text_delta/thinking_delta/input_json_delta paths. The new test file is 336 lines for 119 lines of implementation — excellent ratio. Minor gap: `agent.mcp_tool_use` is routed to `processAgentInputEvent` but has no dedicated test (only `agent.tool_use` and `agent.custom_tool_use` are tested).

### Strengths

- Clean dispatch structure: one function per concern (`processAgentContentEvent`, `processAgentInputEvent`, default passthrough)
- Proper field preservation via raw-map re-serialization — unknown JSON fields survive the round-trip
- Defense-in-depth: unknown `agent.*` events get raw replacement fallback rather than silent passthrough
- `Flush()` before agent event processing prevents accumulator bleed from preceding text_delta chunks
- Tests cover edge cases well (empty content, no-token no-mutation, future event types)

### Verdict

Ready to merge. Two medium/low observations below — neither blocks merge.

---
*Review by Claude*

**Review by laplaque — APPROVED** [review]
## PR Review: feat(anonymizer): handle Managed Agents SSE event types [closes #107]

**Scope:** 455 additions, 0 deletions, 2 files.

### Pipeline status

PASS — all 8 checks green (Lint, Test, Benchmark, Build, Security Scan, CodeQL, Analyze (go), Analyze (actions)).

### Test coverage verdict

PASS — 14 new tests covering all dispatched event types, empty-content passthrough, no-mutation when tokens absent, unknown event fallback, mixed stream interleaving, and regression tests for existing text_delta/thinking_delta/input_json_delta paths. The new test file is 336 lines for 119 lines of implementation — excellent ratio.

### Strengths

- Clean dispatch structure: one function per concern (`processAgentContentEvent`, `processAgentInputEvent`, default passthrough)
- Proper field preservation via raw-map re-serialization — unknown JSON fields survive the round-trip
- Defense-in-depth: unknown `agent.*` events get raw replacement fallback rather than silent passthrough
- `Flush()` before agent event processing prevents accumulator bleed from preceding text_delta chunks
- Tests cover edge cases well (empty content, no-token no-mutation, future event types)

### Non-blocking observations (see inline comments from prior review)

- **MEDIUM:** `processAgentInputEvent` raw-replaces on JSON (same pre-existing pattern as the default case) while `processAgentContentEvent` properly re-serializes. Consider aligning in a follow-up.
- **LOW:** `agent.mcp_tool_use` lacks a dedicated test — only `tool_use` and `custom_tool_use` are tested for the input path.

### Verdict

Approved — clean implementation, solid test coverage, all gates green.

---
*Review by Claude*

File: internal/anonymizer/streaming_anthropic.go (hunk 4 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic.go:271]: **CRITICAL (blocker):** This raw string replacement on `json.RawMessage` can produce invalid JSON when PII originals contain `"`, `\`, or control characters.

Since `processAgentContentEvent` already does the right thing (deserialize → replace in typed fields → re-serialize via `json.Marshal`), this handler should follow the same pattern:

```go
func (a *anthropicDeanonymizer) processAgentInputEvent(payload []byte, agent *agentEventEnvelope) bool {
	if len(agent.Input) == 0 {
		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
		return true
	}

	// Parse input into a generic structure, walk string values, replace tokens.
	var inputObj any
	if err := json.Unmarshal(agent.Input, &inputObj); err != nil {
		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
		return true
	}

	replaced := replaceInValue(inputObj, a.opts.replacer)
	if !replaced {
		writePipe(a.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
		return true
	}

	// Re-serialize preserving all other fields.
	var raw map[string]json.RawMessage
	json.Unmarshal(payload, &raw)
	inputBytes, _ := json.Marshal(inputObj)
	raw["input"] = inputBytes
	out, _ := json.Marshal(raw)
	writePipe(a.opts.pw, []byte(sseDataPrefix), out, []byte("\n"))
	return true
}
```

Where `replaceInValue` recursively walks maps/slices/strings applying the replacer and returns whether any substitution occurred.

**Review by laplaque — CHANGES_REQUESTED** [review]
## Changing to REQUEST_CHANGES

Per review discussion: the raw string replacement in `processAgentInputEvent` is new code introduced in this PR — it's not pre-existing legacy. Since the adjacent `processAgentContentEvent` properly deserializes and re-serializes via `json.Marshal`, the input handler must do the same.

### Blocker

`processAgentInputEvent` (line ~220–240) performs raw `strings.Replace` on `json.RawMessage` content. If the restored PII original contains JSON-special characters (`"`, `\`, newlines), this produces invalid JSON output. Since this is freshly written code (not inherited), it must use proper JSON round-trip semantics — parse the `input` object, walk string values with the replacer, and re-serialize via `json.Marshal`.

Everything else in the PR is solid. Fix this one path and it's good to merge.

---
*Review by Claude*

File: internal/anonymizer/streaming_anthropic_test.go (hunk 5 above)
Comment (laplaque) [inline internal/anonymizer/streaming_anthropic_test.go:297]: **HIGH (blocker):** `agent.mcp_tool_use` is routed to `processAgentInputEvent` but has no test. Every event type that can carry PII tokens must have explicit test coverage — an untested dispatch path is a potential token leak. If this routing ever regresses, nothing catches it.

Add:
```go
func makeAgentMCPToolUse(name string, input map[string]any) string {
	event := map[string]any{
		"type":           "agent.mcp_tool_use",
		"id":             "evt_test_mcptu",
		"name":           name,
		"input":          input,
		"mcp_server_url": "https://example.com/mcp",
		"processed_at":   "2026-05-01T10:00:05Z",
	}
	b, _ := json.Marshal(event)
	return "data: " + string(b) + "\n"
}

func TestAgentMCPToolUseInputReplacement(t *testing.T) {
	token := "[PII_EMAIL_c160f8cc4b2e1a3d]"
	original := "earl@example.com"
	tokenMap := map[string]string{token: original}

	input := map[string]any{"query": "lookup " + token}
	sseInput := makeAgentMCPToolUse("db_query", input)
	got := readStreamResult(t, sseInput+"\n", tokenMap)

	if !strings.Contains(got, original) {
		t.Errorf("agent.mcp_tool_use input token not replaced:\n%s", got)
	}
	if strings.Contains(got, token) {
		t.Errorf("agent.mcp_tool_use still contains token:\n%s", got)
	}
}
```

**Review by laplaque — CHANGES_REQUESTED** [review]
## Additional blocker: missing `agent.mcp_tool_use` test

The dispatch in `processAgentEvent` routes `agent.mcp_tool_use` to `processAgentInputEvent`, but there is no test exercising this path. Untested routing is a potential PII leak — if the dispatch is ever broken or the event type string changes, tokens pass through unreplaced without any test catching it.

Add a `TestAgentMCPToolUseInputReplacement` (with a `makeAgentMCPToolUse` helper) to cover this path. Every event type that carries PII-replaceable content must have its own test.

---
*Review by Claude*

**Comment by iamclaude697** [comment]
All three findings addressed in `56585f3`:

**Finding #1 (HIGH)** — `processAgentContentEvent` now uses `[]json.RawMessage` to parse content blocks individually. Only text blocks with `"type":"text"` get their `text` field patched; all other block types (image, document, tool_use) survive re-serialization with all fields intact. New test `TestAgentMessageMixedContentBlocks` validates this.

**Finding #2 (MEDIUM)** — Comment fixed to reference only reachable event types (`agent.thinking` and unknown `agent.*` events).

**Finding #3 (MEDIUM)** — `TestAgentMessageMixedContentBlocks` added: verifies that an image block with a `source.url` field survives re-serialization when a text block in the same content array triggers replacement.

**Additional fixes:**
- `processAgentInputEvent` rewritten to parse input as JSON object and recursively walk all string values with the replacer (via `replaceStringValues`/`replaceSliceValues`), instead of raw string replacement. Prevents invalid JSON when restored PII contains `"`, `\`, or newlines.
- `TestAgentMCPToolUseInputReplacement` added for the untested `agent.mcp_tool_use` dispatch path.
- `TestAgentToolUseNestedInputReplacement` added for deeply nested input objects.

`make check` passes — 0 lint issues, all tests green, gosec 0 issues, govulncheck clean.

**Review by laplaque — APPROVED** [review]
## Re-review: Both blockers resolved

Both requested changes addressed properly:

1. **JSON-safe input replacement** — `processAgentInputEvent` now uses `replaceStringValues` to recursively walk parsed JSON and re-serializes via `json.Marshal`. Raw replacement only as fallback for non-object inputs. The `processAgentContentEvent` was also improved to use raw `[]json.RawMessage` blocks, preserving unknown fields on non-text block types.

2. **`agent.mcp_tool_use` test coverage** — `TestAgentMCPToolUseInputReplacement` added with `makeAgentMCPToolUse` helper. Additionally: `TestAgentToolUseNestedInputReplacement` confirms recursive replacement in nested objects/arrays, and `TestAgentMessageMixedContentBlocks` verifies image blocks survive content re-serialization.

### Pipeline

All 8 checks green on `56585f3`.

### Verdict

Approved — solid implementation, comprehensive test coverage (423 lines of tests for 194 lines of impl), all gates green.

---
*Review by Claude*
